luacompiler/src/lib/bytecode/mod.rs

```diff
@@ -1,36 +1,160 @@
 pub mod instructions;
 
+use self::instructions::{format_instr, HLInstr};
 use bincode::{deserialize, serialize};
-use irgen::constants_map::ConstantsMap;
+use irgen::{constants_map::ConstantsMap, register_map::Lifetime};
 use std::{
     fmt,
     fs::File,
     io::{self, Write},
     vec::Vec,
 };
 
+/// Represents a function in Lua.
+#[derive(Serialize, Deserialize)]
+pub struct Function<T> {
+    id: usize,
```

Comment:
Why do functions have an ID, and why are there child functions?

Reply:
The compiler has a vector of all the functions it compiled. The ID of a function is its index in that vector.

Each function definition is compiled to a `closure` instruction. The `closure` instruction takes two arguments: a register in which the closure is placed, and the ID of the function which this closure refers to. The problem is that an argument can only address 256 values, therefore a user can only define 256 functions in their program. I realized that each function is compiled as part of another function/module. This means that a function has a parent so to speak. Knowing this, the compiler can assign the index of the child as the second argument of the closure instruction. At run-time this gives enough information to deduce the actual ID of the function to which the closure refers to. This approach allows the compiler to address more than 256 functions.

Reply:
Is `id` being used as a poor-man's way of Rc/Gcing references to functions? Would it not be easier for the compiler to have an `Rc` link to them?

Reply:
The ID is mostly used by the VM. The compiler will emit `closure` instructions for each function, and the VM will associate an ID with the closure. This ensures that when the VM processes a `call` to a closure, it has a way of knowing where to "jump" to, and which instructions to execute.

I think I can get rid of the `parent` for now, and just use `Rc`s to reference the children. Let me see if I can think of two different representations, one which is easier for the compiler to use, and one for the VM. I think I am starting to see what you mean.